### PR TITLE
ir: Use genesis role designations for autodeploy

### PR DESCRIPTION
**Pull request by cthulhu-rider**

### CODE REVIEW

**Comment by cthulhu-rider** [comment]
tested changes against https://github.com/nspcc-dev/neofs-dev-env/pull/282/ and got interested result

--- 

`NeoFSAlphabet` role preset works perfect. Without option:
```
2023-11-17T18:34:03.469Z	info	deploy/deploy.go:303	initial transfer to the committee successfully done
2023-11-17T18:34:03.469Z	info	deploy/deploy.go:305	initializing NeoFS Alphabet...
2023-11-17T18:34:03.469Z	info	deploy/alphabet.go:58	checking NeoFS Alphabet role of the committee members...
2023-11-17T18:34:03.470Z	info	deploy/alphabet.go:80	not all members of the committee have a NeoFS Alphabet role, designation is needed
2023-11-17T18:34:03.471Z	info	deploy/alphabet.go:98	Notary request designating NeoFS Alphabet role to the committee has been successfully sent, will wait for the outcome	{"main tx": "1cb8a7bdc331ea243273816985803f85d77c03b12e5d8ebd7d52d76adf73821b", "fallback tx": "be5261bbf4842c8fee51dc49fdf53ae2cd90821fa00623a38515211eee11079f", "vub": 173}
```
and with:
```
2023-11-17T18:17:27.751Z	info	deploy/deploy.go:303	initial transfer to the committee successfully done
2023-11-17T18:17:27.751Z	info	deploy/deploy.go:305	initializing NeoFS Alphabet...
2023-11-17T18:17:27.751Z	info	deploy/alphabet.go:58	checking NeoFS Alphabet role of the committee members...
2023-11-17T18:17:27.752Z	info	deploy/alphabet.go:76	all committee members have a NeoFS Alphabet role
2023-11-17T18:17:27.752Z	info	deploy/deploy.go:318	NeoFS Alphabet successfully initialized
```

---

`P2PNotary` role works too. Without option:
```
2023-11-17T18:33:59.450Z	info	deploy/deploy.go:255	NNS contract successfully initialized on the chain	{"address": "2e9a67b9a56562541add67af6056b49b486d67b0"}
2023-11-17T18:33:59.450Z	info	deploy/deploy.go:257	enable Notary service for the committee...
2023-11-17T18:33:59.450Z	info	deploy/notary.go:66	committee is single-acc, no multi-signature needed for Notary role designation
2023-11-17T18:33:59.450Z	info	deploy/notary.go:97	checking Notary role of the committee members...
2023-11-17T18:33:59.450Z	info	deploy/notary.go:119	not all members of the committee have a Notary role, designation is needed
2023-11-17T18:33:59.450Z	info	deploy/notary.go:152	sending new transaction designating Notary role to the local account...
2023-11-17T18:33:59.452Z	info	deploy/notary.go:166	transaction designating Notary role to the local account has been successfully sent, will wait for the outcome	{"tx": "1d3d0a871033b4e38c95e1ee6c85149925f96e64e96ab572cfb3efcc76ab1bf2", "vub": 29}
```
and with
```
2023-11-17T18:28:05.211Z	info	deploy/deploy.go:255	NNS contract successfully initialized on the chain	{"address": "2e9a67b9a56562541add67af6056b49b486d67b0"}
2023-11-17T18:28:05.211Z	info	deploy/deploy.go:257	enable Notary service for the committee...
2023-11-17T18:28:05.211Z	info	deploy/notary.go:66	committee is single-acc, no multi-signature needed for Notary role designation
2023-11-17T18:28:05.211Z	info	deploy/notary.go:97	checking Notary role of the committee members...
2023-11-17T18:28:05.212Z	info	deploy/notary.go:115	all committee members have a Notary role
2023-11-17T18:28:05.212Z	info	deploy/deploy.go:273	Notary service successfully enabled for the committee
```

---

so, everything seems to be OK.. but initial transfer breaks with `P2PNotary` role preset. In test env there is only one node that is committee/validators/Alphabet. When `set_roles_in_genesis` is set, https://github.com/nspcc-dev/neofs-node/blob/8d9a2bccdedfeb075c68df6a1d8382952baef37d/pkg/morph/deploy/funds.go#L59 hangs.

i compared transfer transactions with (when notary request with tx is accepted) and without (hang case) the option. Without:
```
2023-11-17T18:34:02.464Z	info	deploy/funds.go:121	got current balance of the validator multi-sig account, distributing between the committee...	{"NEO": "100000000", "GAS": "5199999993361180"}
2023-11-17T18:34:02.464Z	info	deploy/funds.go:152	not enough GAS on the committee member's account, need replenishment	{"member": "b3622bf4017bdfe317c58aed5f4c753f206b7db896046fa7d774bbc4bf7f8dc2af9c7b29759df7f3d92052a5b9bc545bcd31c6a7a3463e90c768a6c3e45b1036", "balance": "148374010", "need at least": "15000000000"}
2023-11-17T18:34:02.465Z	info	deploy/funds.go:219	sending new Notary request transferring funds from validator multi-sig account to the committee...
2023-11-17T18:34:02.468Z	info	deploy/funds.go:231	Notary request transferring funds from validator multi-sig account to the committee has been successfully sent, will wait for the outcome	{"main tx": "80c55919db52e69aca03b66ccc5318000663958694b45ff4213c5cf03329339f", "fallback tx": "557e43423ed8a253245837280e0336122b5e4cdd489f19972c731513f692c087", "vub": 172}
2023-11-17T18:34:02.469Z	info	deploy/notary.go:1128	main transaction from the received notary request already has local account's signature, skip	{"tx": "80c55919db52e69aca03b66ccc5318000663958694b45ff4213c5cf03329339f"}
2023-11-17T18:34:03.191Z	info	deploy/notary.go:1096	main transaction of the notary request has already been processed, skip	{"tx": "80c55919db52e69aca03b66ccc5318000663958694b45ff4213c5cf03329339f"}
2023-11-17T18:34:03.191Z	info	deploy/util.go:88	new block arrived	{"height": 30}
2023-11-17T18:34:03.192Z	info	deploy/notary.go:980	enough funds on the notary balance, deposit is not needed	{"balance": "100000000"}
2023-11-17T18:34:03.469Z	info	deploy/funds.go:121	got current balance of the validator multi-sig account, distributing between the committee...	{"NEO": "100000000", "GAS": "5199969993361180"}
2023-11-17T18:34:03.469Z	info	deploy/funds.go:146	enough GAS on the committee member's account, skip transfer	{"member": "b3622bf4017bdfe317c58aed5f4c753f206b7db896046fa7d774bbc4bf7f8dc2af9c7b29759df7f3d92052a5b9bc545bcd31c6a7a3463e90c768a6c3e45b1036", "balance": "30188396230", "need at least": "15000000000"}
```
and with:
```
2023-11-17T18:28:07.223Z	info	deploy/funds.go:152	not enough GAS on the committee member's account, need replenishment	{"member": "b3622bf4017bdfe317c58aed5f4c753f206b7db896046fa7d774bbc4bf7f8dc2af9c7b29759df7f3d92052a5b9bc545bcd31c6a7a3463e90c768a6c3e45b1036", "balance": "84892290", "need at least": "15000000000"}
2023-11-17T18:28:07.223Z	info	deploy/funds.go:219	sending new Notary request transferring funds from validator multi-sig account to the committee...
2023-11-17T18:28:07.226Z	info	deploy/funds.go:231	Notary request transferring funds from validator multi-sig account to the committee has been successfully sent, will wait for the outcome	{"main tx": "4caaca85ddb5d8046e69057520e11def7c50fc82e96f0811fbc4a239af385959", "fallback tx": "dee0ce208be68a6dd41adf87f2207eac333e52cb74aa897d8b1668d26cb05eec", "vub": 171}
2023-11-17T18:28:07.227Z	info	deploy/notary.go:1128	main transaction from the received notary request already has local account's signature, skip	{"tx": "4caaca85ddb5d8046e69057520e11def7c50fc82e96f0811fbc4a239af385959"}
```

i checked tx scripts: in both cases they are same. The only observable diff from logs is that no-option case writes two messages `main transaction from the received notary request already has local account's signature, skip` and `main transaction of the notary request has already been processed, skip` while with-option case writes only 1st message. Accordingly, when option is unset, Notary request goes to the channel twice (and TX succeeds), and once - when option is set (and TX is lost)
 
overall, for now i have no idea why process stucks. Need more deep research with fresh head. Help is appreciated

**Comment by cthulhu-rider** [comment]
tested changes against https://github.com/nspcc-dev/neofs-dev-env/pull/282/ and got interested result

--- 

`NeoFSAlphabet` role preset works perfect. Without option:
```
2023-11-17T18:34:03.469Z	info	deploy/deploy.go:303	initial transfer to the committee successfully done
2023-11-17T18:34:03.469Z	info	deploy/deploy.go:305	initializing NeoFS Alphabet...
2023-11-17T18:34:03.469Z	info	deploy/alphabet.go:58	checking NeoFS Alphabet role of the committee members...
2023-11-17T18:34:03.470Z	info	deploy/alphabet.go:80	not all members of the committee have a NeoFS Alphabet role, designation is needed
2023-11-17T18:34:03.471Z	info	deploy/alphabet.go:98	Notary request designating NeoFS Alphabet role to the committee has been successfully sent, will wait for the outcome	{"main tx": "1cb8a7bdc331ea243273816985803f85d77c03b12e5d8ebd7d52d76adf73821b", "fallback tx": "be5261bbf4842c8fee51dc49fdf53ae2cd90821fa00623a38515211eee11079f", "vub": 173}
```
and with:
```
2023-11-17T18:17:27.751Z	info	deploy/deploy.go:303	initial transfer to the committee successfully done
2023-11-17T18:17:27.751Z	info	deploy/deploy.go:305	initializing NeoFS Alphabet...
2023-11-17T18:17:27.751Z	info	deploy/alphabet.go:58	checking NeoFS Alphabet role of the committee members...
2023-11-17T18:17:27.752Z	info	deploy/alphabet.go:76	all committee members have a NeoFS Alphabet role
2023-11-17T18:17:27.752Z	info	deploy/deploy.go:318	NeoFS Alphabet successfully initialized
```

---

`P2PNotary` role works too. Without option:
```
2023-11-17T18:33:59.450Z	info	deploy/deploy.go:255	NNS contract successfully initialized on the chain	{"address": "2e9a67b9a56562541add67af6056b49b486d67b0"}
2023-11-17T18:33:59.450Z	info	deploy/deploy.go:257	enable Notary service for the committee...
2023-11-17T18:33:59.450Z	info	deploy/notary.go:66	committee is single-acc, no multi-signature needed for Notary role designation
2023-11-17T18:33:59.450Z	info	deploy/notary.go:97	checking Notary role of the committee members...
2023-11-17T18:33:59.450Z	info	deploy/notary.go:119	not all members of the committee have a Notary role, designation is needed
2023-11-17T18:33:59.450Z	info	deploy/notary.go:152	sending new transaction designating Notary role to the local account...
2023-11-17T18:33:59.452Z	info	deploy/notary.go:166	transaction designating Notary role to the local account has been successfully sent, will wait for the outcome	{"tx": "1d3d0a871033b4e38c95e1ee6c85149925f96e64e96ab572cfb3efcc76ab1bf2", "vub": 29}
```
and with
```
2023-11-17T18:28:05.211Z	info	deploy/deploy.go:255	NNS contract successfully initialized on the chain	{"address": "2e9a67b9a56562541add67af6056b49b486d67b0"}
2023-11-17T18:28:05.211Z	info	deploy/deploy.go:257	enable Notary service for the committee...
2023-11-17T18:28:05.211Z	info	deploy/notary.go:66	committee is single-acc, no multi-signature needed for Notary role designation
2023-11-17T18:28:05.211Z	info	deploy/notary.go:97	checking Notary role of the committee members...
2023-11-17T18:28:05.212Z	info	deploy/notary.go:115	all committee members have a Notary role
2023-11-17T18:28:05.212Z	info	deploy/deploy.go:273	Notary service successfully enabled for the committee
```

---

so, everything seems to be OK.. but initial transfer breaks with `P2PNotary` role preset. In test env there is only one node that is committee/validators/Alphabet. When `set_roles_in_genesis` is set, https://github.com/nspcc-dev/neofs-node/blob/8d9a2bccdedfeb075c68df6a1d8382952baef37d/pkg/morph/deploy/funds.go#L59 hangs.

i compared transfer transactions with (when notary request with tx is accepted) and without (hang case) the option. Without:
```
2023-11-17T18:34:02.464Z	info	deploy/funds.go:121	got current balance of the validator multi-sig account, distributing between the committee...	{"NEO": "100000000", "GAS": "5199999993361180"}
2023-11-17T18:34:02.464Z	info	deploy/funds.go:152	not enough GAS on the committee member's account, need replenishment	{"member": "b3622bf4017bdfe317c58aed5f4c753f206b7db896046fa7d774bbc4bf7f8dc2af9c7b29759df7f3d92052a5b9bc545bcd31c6a7a3463e90c768a6c3e45b1036", "balance": "148374010", "need at least": "15000000000"}
2023-11-17T18:34:02.465Z	info	deploy/funds.go:219	sending new Notary request transferring funds from validator multi-sig account to the committee...
2023-11-17T18:34:02.468Z	info	deploy/funds.go:231	Notary request transferring funds from validator multi-sig account to the committee has been successfully sent, will wait for the outcome	{"main tx": "80c55919db52e69aca03b66ccc5318000663958694b45ff4213c5cf03329339f", "fallback tx": "557e43423ed8a253245837280e0336122b5e4cdd489f19972c731513f692c087", "vub": 172}
2023-11-17T18:34:02.469Z	info	deploy/notary.go:1128	main transaction from the received notary request already has local account's signature, skip	{"tx": "80c55919db52e69aca03b66ccc5318000663958694b45ff4213c5cf03329339f"}
2023-11-17T18:34:03.191Z	info	deploy/notary.go:1096	main transaction of the notary request has already been processed, skip	{"tx": "80c55919db52e69aca03b66ccc5318000663958694b45ff4213c5cf03329339f"}
2023-11-17T18:34:03.191Z	info	deploy/util.go:88	new block arrived	{"height": 30}
2023-11-17T18:34:03.192Z	info	deploy/notary.go:980	enough funds on the notary balance, deposit is not needed	{"balance": "100000000"}
2023-11-17T18:34:03.469Z	info	deploy/funds.go:121	got current balance of the validator multi-sig account, distributing between the committee...	{"NEO": "100000000", "GAS": "5199969993361180"}
2023-11-17T18:34:03.469Z	info	deploy/funds.go:146	enough GAS on the committee member's account, skip transfer	{"member": "b3622bf4017bdfe317c58aed5f4c753f206b7db896046fa7d774bbc4bf7f8dc2af9c7b29759df7f3d92052a5b9bc545bcd31c6a7a3463e90c768a6c3e45b1036", "balance": "30188396230", "need at least": "15000000000"}
```
and with:
```
2023-11-17T18:28:07.223Z	info	deploy/funds.go:152	not enough GAS on the committee member's account, need replenishment	{"member": "b3622bf4017bdfe317c58aed5f4c753f206b7db896046fa7d774bbc4bf7f8dc2af9c7b29759df7f3d92052a5b9bc545bcd31c6a7a3463e90c768a6c3e45b1036", "balance": "84892290", "need at least": "15000000000"}
2023-11-17T18:28:07.223Z	info	deploy/funds.go:219	sending new Notary request transferring funds from validator multi-sig account to the committee...
2023-11-17T18:28:07.226Z	info	deploy/funds.go:231	Notary request transferring funds from validator multi-sig account to the committee has been successfully sent, will wait for the outcome	{"main tx": "4caaca85ddb5d8046e69057520e11def7c50fc82e96f0811fbc4a239af385959", "fallback tx": "dee0ce208be68a6dd41adf87f2207eac333e52cb74aa897d8b1668d26cb05eec", "vub": 171}
2023-11-17T18:28:07.227Z	info	deploy/notary.go:1128	main transaction from the received notary request already has local account's signature, skip	{"tx": "4caaca85ddb5d8046e69057520e11def7c50fc82e96f0811fbc4a239af385959"}
```

i checked tx scripts: in both cases they are same. The only observable diff from logs is that no-option case writes two messages `main transaction from the received notary request already has local account's signature, skip` and `main transaction of the notary request has already been processed, skip` while with-option case writes only 1st message. Accordingly, when option is unset, Notary request goes to the channel twice (and TX succeeds), and once - when option is set (and TX is lost)
 
overall, for now i have no idea why process stucks. Need more deep research with fresh head. Help is appreciated

**Comment by cthulhu-rider** [comment]
this is probably related to Notary service non-start, figuring this out with @AnnaShaleva and @roman-khimov ...

**Comment by cthulhu-rider** [comment]
works after [fix](https://github.com/nspcc-dev/neo-go/pull/3229) ([revision](7a1bf77585a37302ddfae78b2c88ed472f66cbcc)). Lets wait for it to be released

File: pkg/innerring/config.go [hunk 1]
```diff
@@ -113,15 +113,20 @@ func parseBlockchainConfig(v *viper.Viper, _logger *zap.Logger) (c blockchain.Co
 			num, err := cast.ToUint32E(val)
 			if err != nil {
 				return err
-			} else if num > math.MaxInt32 {
+			} else if num <= 0 {
 				return fmt.Errorf("value %d is out of allowable range", num)
+			} else if num > uint32(c.Committee.Len()) {
```
[the remark this answers is not in the document]
Comment (cthulhu-rider) [reply pkg/innerring/config.go:118]: not this issue's point but interesting, moved to #2666

File: pkg/innerring/internal/blockchain/blockchain.go [hunk 2]
```diff
@@ -336,13 +344,26 @@ func New(cfg Config) (res *Blockchain, err error) {
...
 	cfgBaseProto.Hardforks = cfg.HardForks
+	var genesisValidatorsCount uint32
 	if cfg.ValidatorsHistory != nil {
 		cfgBaseProto.ValidatorsHistory = make(map[uint32]uint32, len(cfg.ValidatorsHistory))
 		for height, num := range cfg.ValidatorsHistory {
 			cfgBaseProto.ValidatorsHistory[height] = num
 		}
+
+		// value checked above in function
+		genesisValidatorsCount = cfg.ValidatorsHistory[0]
```
[the remark this answers is not in the document]
Comment (cthulhu-rider) [reply pkg/innerring/internal/blockchain/blockchain.go]: used

**Review by roman-khimov — APPROVED** [review]

**Comment by roman-khimov** [comment]
Conflicts though.

File: pkg/innerring/config.go [hunk 3]
```diff
@@ -381,7 +389,7 @@ func parseConfigMap(v *viper.Viper, key, desc string, f func(name string, val an
 
 func parseConfigMapUint32(v *viper.Viper, key, desc string, limit uint64) (map[string]uint32, error) {
 	res := make(map[string]uint32)
-	return res, parseConfigMap(v, key, desc, func(name string, val any) error {
+	return res, parseConfigMap(v, key, desc, func(name string, val interface{}) error {
```
[the remark this answers is not in the document]
Comment (cthulhu-rider) [reply pkg/innerring/config.go]: fixed

**Pull request merged**